### ISSUE TITLE
toBlob() Having two optional kinda sucks... let's merge them.

### ISSUE THREAD

**Issue by marcoscaceres** [WICG/img-conversion#13]
right now: 

``` IDL
  Promise<Blob> toBlob (optional DOMString type, optional EncoderOpts options);
```

Why not just add the `type` member to `EncoderOpts`? It will work the same as currently specified, as when missing, just assume type="image/png".

**Comment by marcoscaceres** [comment]
Additionally, EncoderOpts should be renamed to `ImageOps` or something so it can be used in encoding and decoding.

**Comment by AshleyScirra** [comment]
Canvas' toBlob is already specced like this, and this method is intended to be almost equivalent, so I thought it would be best to remain consistent. Or, if you want to make this change, it should also be made to the canvas version of toBlob. Either way IMO both should support the same way instead of diverging.

**Comment by marcoscaceres** [comment]
I'm not very concerned about following canvas. Canvas is one of the worst designed APIs in the platform.

**Comment by AshleyScirra** [comment]
Are you referring to the HTMLCanvasElement interface where toBlob is defined? That is pretty much in line with normal DOM elements, what do you think is so bad about it? Maybe you have a poor opinion of CanvasRenderingContext2D or WebGL, but they are separate APIs to HTMLCanvasElement - neither defines a toBlob method. As a web developer I would favour consistency in the platform I work with unless there is some material benefit to making the change, and the only consequence of tweaking something like the "type" parameter of toBlob is to make developers confused about which version of toBlob they are using.

**Comment by marcoscaceres** [comment]
Sorry, you are correct. It's CanvasRenderingContext2D that I, and many others, hate. 

I guess the only other API I can think of the top of my head is Fetch's Request.blob() - which returns a promise: https://fetch.spec.whatwg.org/#dom-body-blob

I generally agree about consistency, but are we not already modifying the signature of .toBlob()  by  dropping the callback as the first argument?

**Comment by AshleyScirra** [comment]
Yes, but I expect the Canvas toBlob to be similarly overloaded, as part of the overall movement of the web platform towards promises over callbacks. I specced ImageData's toBlob to return a Promise in anticipation of this, but I would be happy to include both the callback and promise form for complete consistently with canvas toBlob.

The Request blob() method does use a slightly different name. I think we might be bikeshedding over this, but I feel it's OK to use different names, since toBlob() is used for a conversion from one thing to another, whereas blob() is used to obtain the request body in the first place.

**Comment by marcoscaceres** [comment]
>  I specced ImageData's toBlob to return a Promise in anticipation of this, but I would be happy to include both the callback and promise form for complete consistently with canvas toBlob.

I would not be in favor of that. That would just be confusing. 

> The Request blob() method does use a slightly different name. I think we might be bikeshedding over this, but I feel it's OK to use different names, since toBlob() is used for a conversion from one thing to another, whereas blob() is used to obtain the request body in the first place.

The thing is, conceptually, just like fetch, we are requesting an operation to happen to something over time. The returned representation is really an operation that whose result may be a blob. However, what is being constructed is a stream that eventually results in a blob. 

This leads me to believe that returning a promise is wrong. We really want to return either a pure stream, or a sub-class of stream that you can call .blob() on, that returns a promise.

**Comment by marcoscaceres** [comment]
To be clear:

``` JS
var stream =  imageData.doThing(options);
steam.pipe(steamingUploader); // upload as encode/decode 
showItWhenDone(await stream.blob()); //promise to resolve to a blob

function showItWhenDone(blob){
   someImg.src = URL.createObjectURL(blob); 
}
```

**Comment by marcoscaceres** [comment]
Updated example above.

**Comment by AshleyScirra** [comment]
I am not really familiar with the upcoming streams API, but it sounds like a good idea. I'm wondering if there's much performance benefit though, since image encoding/decoding is typically pretty fast (on the order of 100ms). So in your upload example it seems just as practical to wait for doThing() to complete in its entirety before uploading, since the network part would probably take orders of magnitude longer. Perhaps streams could wait for a v2 spec?

**Comment by marcoscaceres** [comment]
@AshleyScirra I think we might want it for v1, so that we can do things like progress as well (e.g., commutative encoding/decoding of multiple images, etc) - one could also be applying effects chunks of the stream come in, etc.

@jrmuizel, any additional use cases you might like to add re: stream? How do you envision them being used?
